# Examen de prácticas. Curso 22-23. Convocatoria Enero 2023

## Aplicación de la factorización SVD a la compresión de imágenes

En este ejercicio veremos cómo se puede comprimir una imagen .jpeg haciendo uso de la Descomposición en Valores Singulares (SVD).


En la carpeta **data** contrarás el fichero **beltrami_adult.npy**, el cual almacena un array de numpy. Has de hacer los siguiente:

1) Carga dicho fichero (recuerda que esto lo hicimos al final de la práctica dedicada a tensores) y guárdalo en una variable que llamaremos $M$.

2) Cálcula la forma de dicho array

3) Transforma el array que has cargado a una imagen del módulo PIL (revisa la práctica de tensores). Llama **beltrami** a la imagen resultante.

4) Muestra por pantalla dicha imagen. Para ello, escribe **beltrami.show()**

**Puntos = 0.5**



In [1]:
# Completar aquí

import numpy as np

from PIL import Image

M = np.load("../data/beltrami_adult.npy")
print(f"forma del array =  {M.shape}")

beltrami = Image.fromarray(M)

beltrami.show()

# Fin Completar aquí ------------------------------------


forma del array =  (228, 187)


Como puedes imaginar, la foto que ves en pantalla es de un matemático. En concreto, se trata del matemático italiano Eugenio Beltrami (1835 – 1900), uno de los precursores de la descomposición SVD (aunque en un contexto que nada tiene que ver con la Ciencia de Datos) .

Como puedes ver, Python ha almacenado dicha imagen en una matriz $M$ de tamaño $228\times 187$. 

Como el número de filas es mayor que el de columnas, por razones técnicas que no interesan de momento, calcula la traspuesta de $M$ y guardarla en una matriz que llamaremos $A$.

Para asegurarte, imprime la forma de la matriz $A$.

Respecto de la imagen de Beltrami, esto supone girar dicha imagen. 

**Puntos = 0.25**



In [2]:
# Completar aquí

A = M.T

print(f"forma de A = {A.shape}")

# Fin Completar aquí ------------------------------------

forma de A = (187, 228)


Vamos a utilizar la descomposición en valores singulares (SVD) para comprimir dicha imagen.

Así pues, calcula la factorización SVD de la matriz anterior $A$, es decir, encuentra matrices $U$, $\Sigma$ y $V$, de modo que $A = U\Sigma V^T$. Calcula la forma de estas tres matrices.

**Puntos = 4.25**

In [3]:
# Completar aquí

from scipy.linalg import svd

U, s, V = svd(A)

# creamos una matriz de zeros de tamaño m x n 
Sigma = np.zeros((A.shape[0], A.shape[1]))

# rellenamos Sigma con las valores singulares
Sigma[:A.shape[0], :A.shape[0]] = np.diag(s)

print(f"U = \n {U}")
print(f"S = \n {Sigma}")
print(f"V^T = {V}")

print(f"shape of U = {U.shape}")
print(f"shape of Sigma = {Sigma.shape}")
print(f"shape of V = {V.shape}")

# Fin Completar aquí ------------------------------------

U = 
 [[-0.10358145  0.02198968  0.0021542  ... -0.0759327  -0.00146994
  -0.0177579 ]
 [-0.09844697  0.01901818  0.00894816 ...  0.2003558   0.11289407
  -0.01032199]
 [-0.0951115   0.01812917  0.0152985  ... -0.22930472 -0.15069006
   0.02101575]
 ...
 [-0.11506867  0.03656439 -0.054218   ... -0.17008455  0.07608643
   0.3568638 ]
 [-0.1167452   0.03590726 -0.05989129 ...  0.11246515  0.08630929
   0.1140729 ]
 [-0.11755426  0.03489203 -0.06180156 ...  0.08784679 -0.13992873
  -0.44077307]]
S = 
 [[29036.01367188     0.             0.         ...     0.
      0.             0.        ]
 [    0.          6431.21728516     0.         ...     0.
      0.             0.        ]
 [    0.             0.          4946.71435547 ...     0.
      0.             0.        ]
 ...
 [    0.             0.             0.         ...     0.
      0.             0.        ]
 [    0.             0.             0.         ...     0.
      0.             0.        ]
 [    0.             0.             

Usando la técnica de **slicing** vamos a quedarnos con los primeros $r$ mayores valores singulares de $\Sigma$. Para ello:

1) Introduce una variable, llamada $r$, con valor $r = 50$.

Utilizando la técnica de slicing: 

2) Extrae de $U$ las primeras $r$ columnas. Denota por $U_r$ la matriz resultante. Imprime la forma de $U_r$.

3) Genera la matriz $\Sigma_r$ extrayendo las primeras $r$ filas y las primeras $r$ columnas de $\Sigma$. Imprime la forma de $\Sigma_r$.

4) Extrae de $V$ las primeras $r$ filas. Denota por $V_r$ la matriz resultante. Imprime la forma de $V_r$.


**Puntos = 4.25**

In [4]:
# Completar aquí

r = 50
U_r = U[:, :r]
print(f"shape de U_r = {U_r.shape}")
Sigma_r = Sigma[:r, :r]
print(f"shape de Sigma_r = {Sigma_r.shape}")
V_r = V[:r, :]
print(f"shape de U_r = {V_r.shape}")

# Fin Completar aquí ------------------------------------

shape de U_r = (187, 50)
shape de Sigma_r = (50, 50)
shape de U_r = (50, 228)


Genera la matriz $A_r = U_r\Sigma_r V_r$ e imprime su forma.

**Puntos = 0.5**


In [5]:
# Completar aquí

A_r = U_r @ Sigma_r @ V_r
print(f"forma de A_r = {A_r.shape}")

# Fin Completar aquí ------------------------------------

forma de A_r = (187, 228)


Calcula la traspuesta de $A_r$ y denótala por $M_r$.

Transforma $M_r$ en una imagen del módulo PIL y llámala beltrami_compressed

Podrás ver la imagen comprimida escribiendo **beltrami_compressed.show()**

**Puntos = 0.25**

In [6]:
# Completar aquí

M_r = A_r.T

beltrami_compressed = Image.fromarray(M_r)

beltrami_compressed.show()

# Fin Completar aquí ------------------------------------

Verás que apenas se nota diferencia con la imagen inicial. Si quieres apreciar claramente el efecto de la compresión, pon $r=10$ y ejecuta de nuevo todo el código.

Esto es todo. Pero, volvamos a la parte de **compresión de imágenes** que aparece en el título de esta práctica.

Cuando una imagen aparece guardada en una matriz, como la matriz $A$ de partida, se dice que **no está comprimida**. Para almacenar dicha matriz hemos de guardar $m\times n$ datos. En el caso de $A$, esto nos da $187\times 228 = 42636$ datos. Se dice que la imagen está **comprimida** si está almacenada en formato SVD. En nuestro ejemplo, $A_r = U_r\Sigma_r V_r^T$, necesitamos guardar $m\times r + r + r\times n $ datos, que para $r = 50$ nos da $20800$. Podemos interpretar estos datos de una manera muy sencilla: **la imagen comprimida necesita almacenar aproximadamente la mitad de datos que sin comprimir.**

Espero que este ejercicio haya servido para empezar a  apreciar el destacadísimo papel de la SVD en Ciencia de Datos.   